### ISSUE TITLE
False positive in `SwitchDefault` on Java 21 pattern matching.

### ISSUE THREAD

**Issue by oxkitsune** [google/error-prone#4266]
There's a false positive in the `SwitchDefault` checker when using Java 21's pattern matching in switch cases.

Simple example
```java
sealed interface Foo {
  final class Bar implements Foo {}
}

class Test {
  void f(Foo i) {
    switch (i) {
      case Foo.Bar bar -> {}
    }
  }
}
```

This results in: 
```
warning: [SwitchDefault] The default case of a switch should appear at the end of the last statement group
            case Foo.Bar bar -> {}
            ^
    (see https://errorprone.info/bugpattern/SwitchDefault)
  Did you mean to remove this line?
```

This happens because of the way default cases are detected in the `SwitchDefault` checker at [Line 47](https://github.com/google/error-prone/blob/ec9d69a8bcaa783e494a066351fe23c21640330b/core/src/main/java/com/google/errorprone/bugpatterns/SwitchDefault.java#L47). `CaseTree#getExpression()`  will return `null` for the `default` case, but also for the new `CaseLabelTree` cases that are used for the pattern matching.

In JDK21 there's `CaseTree#getLabels()` which will return a list containing a `DefaultCaseLabelTree`, which could be used to differentiate between these cases. 
I haven't been able to find a reliable method to differentiate between these two without relying on JDK21 features. Perhaps we could do some naive string checks? E.g. 
```java
Optional<? extends CaseTree> maybeDefault =
    tree.getCases().stream()
        .filter(c -> state.getSourceForNode(c).contains("default"))
        .findAny();
```
This example will need to handle the different syntax for `CaseKind.RULE` and `CaseKind.STATEMENT` ofcourse.
